File: advent-of-code/2020/Day11.scala
import scala.io.StdIn.{readLine}
import scala.util.control.Breaks.{break, breakable}

object Status extends Enumeration {
  type Status = Value
  val FLOOR, EMPTY, OCCUPIED = Value
}

object Day11 {

  import Status._

  def mutate[T](grid: Vector[Vector[T]], step: (Vector[Vector[T]], T, Int, Int) => T): Int = {
    var gridCopy = grid

    breakable {
      while (true) {
        val newSeats = gridCopy.toSeq.zipWithIndex.map { case (row, i) =>
          row.toSeq.zipWithIndex.map { case (seat, j) => step(gridCopy, seat, i, j) }
        }
        if (newSeats == gridCopy) break()
        gridCopy = newSeats
      }
    }

    return gridCopy.toSeq.map(_.count(_ == Status.OCCUPIED)).sum
  }

  def step(status: Status, adjacent: Vector[Status]) =
    if (status == Status.EMPTY && adjacent.forall(_ != Status.OCCUPIED)) Status.OCCUPIED
    else if (status == Status.OCCUPIED && adjacent.count(_ == Status.OCCUPIED) >= 5) Status.EMPTY
    else status

  def neighbours[T](element: Vector[T], i: Int, f: (Seq[T]) => Seq[Status]) =
    Vector(
      f(element.toSeq).take(i).findLast(it => it != null && it != Status.FLOOR),
      f(element.toSeq).drop(i + 1).find(it => it != null && it != Status.FLOOR)
    ).filter(_ != null)

  def main(args: Array[String]) = {
    var seats = Vector[Vector[Status]]()

    var line = readLine()
    while (line != null) {
      seats = seats :+ line.map(it => if (it == 'L') EMPTY else FLOOR).toVector
      line = readLine()
    }

    val first = mutate(
      seats,
      { (grid: Vector[Vector[Status]], seat: Status, i: Int, j: Int) =>
        val adjacent = grid
          .drop(math.max(i - 1, 0))
          .dropRight(grid.size - i - 2)
          .flatMap(it => it.drop(math.max(j - 1, 0)).dropRight(it.size - j - 2))
        step(seat, adjacent)
      }
    )

    val second = mutate(
      seats,
      { (grid: Vector[Vector[Status]], seat: Status, i: Int, j: Int) =>
        val adjacent = Vector(
          neighbours(grid.apply(i), j, (it: Seq[Status]) => it),
          neighbours(grid, i, (it: Seq[Vector[Status]]) => it.map(_.apply(j))),
          neighbours(grid, i, (it: Seq[Vector[Status]]) =>
            it.toSeq.zipWithIndex.map { case (s, index) => {
              val id = index - i + j
              if (id >= 0 && id < s.size) s.apply(id) else null
            }}
          ),
          neighbours(grid, i, (it: Seq[Vector[Status]]) =>
            it.toSeq.zipWithIndex.map { case (s, index) => {
              val id = -index + i + j
              if (id >= 0 && id < s.size) s.apply(id) else null
            }}
          )
        )
          .flatMap(it => it)
          .collect { case Some(i) => i }

        step(seat, adjacent)
      }
    )

    println("%d %d".format(first, second))
  }
}

File: advent-of-code/2020/Day11.kt
import java.lang.Integer.max
import java.lang.Integer.min

enum class Status {
    FLOOR, EMPTY, OCCUPIED
}

fun <T> mutate(grid: List<List<T>>, step: (List<List<T>>, T, Int, Int) -> T): Int {
    var gridCopy = grid

    while (true) {
        val newSeats = gridCopy.mapIndexed { i, row -> row.mapIndexed { j, seat -> step(gridCopy, seat, i, j) } }
        if (newSeats == gridCopy) break
        gridCopy = newSeats
    }

    return gridCopy.sumBy { it.count { it == Status.OCCUPIED } }
}

fun step(status: Status, adjacent: List<Status>) =
    when {
        status == Status.EMPTY && adjacent.all { it != Status.OCCUPIED } -> Status.OCCUPIED
        status == Status.OCCUPIED && adjacent.count { it == Status.OCCUPIED } >= 5 -> Status.EMPTY
        else -> status
    }

fun <T> neighbours(element: Collection<T>, i: Int, f: (Sequence<T>) -> Sequence<Status?>) =
    sequenceOf(
        element.asSequence().run(f).take(i).findLast { it != null && it != Status.FLOOR },
        element.asSequence().run(f).drop(i + 1).find { it != null && it != Status.FLOOR }
    )
    .mapNotNull { it }

fun main() {
    val seats = mapLines { it.map { if (it == 'L') Status.EMPTY else Status.FLOOR } }

    val first = mutate(seats) { grid, seat, i, j ->
        val adjacent = grid.slice(max(i - 1, 0)..min(i + 1, grid.lastIndex))
            .flatMap { it.slice(max(j - 1, 0)..min(j + 1, it.lastIndex)) }
        step(seat, adjacent)
    }

    val second = mutate(seats) { grid, seat, i, j ->
        if (seat == Status.FLOOR) return@mutate seat

        val adjacent = sequenceOf(
                neighbours(grid[i], j) { it },
                neighbours(grid, i) { it.map { it[j] } },
                neighbours(grid, i) { it.mapIndexed { index, s -> s.getOrNull(index - i + j) } },
                neighbours(grid, i) { it.mapIndexed { index, s -> s.getOrNull(-index + i + j) } }
            )
            .flatMap { it }
            .toList()
        step(seat, adjacent)
    }

    println("$first $second")
}
